modules/xmpp/JingleSessionPC.js

```diff
@@ -1196,203 +1398,248 @@ export default class JingleSessionPC extends JingleSession {
+     * is removed from this session and the renegotiation is performed.
+     * The promise will be rejected with a <tt>string</tt> that the describes
+     * the error if anything goes wrong.
+     */
+    removeTrackAsMute(track) {
+        if (!track) {
+            return Promise.reject('invalid "stream" argument value');
         }
-
-        // Find the right sender (for audio or video)
-        this.peerconnection.peerconnection.getSenders().some(s => {
-            if (s.track === track) {
-                sender = s;
-
-                return true;
+        const workFunction = () => {
+            if (!this.peerconnection) {
+                return false;
             }
 
-            return false;
-        });
+            return this.peerconnection.removeTrackMute(track);
+        };
 
-        if (sender) {
-            this.peerconnection.peerconnection.removeTrack(sender);
-        } else {
-            logger.log('Cannot remove tracks: no RTPSender.');
-        }
+        return this._doRenegotiate('remove-as-mute', workFunction);
     }
 
     /**
-     * Just remove the stream from the peerconnection
-     * @param {JitsiLocalTrack|MediaStream} stream the stream to remove
-     * NOTE: must be called within a work function being executed
-     *  by the modification queue.
+     * A worker function that is scheduled and executed on the
+     * {@link modificationQueue}. Local SDPs from before and after the function
+     * is executed are compared and 'source-add'/'source-remove' notifications
+     * are being sent to remote participants (to propagate changes in local
+     * streams description).
+     * @name JingleSessionPC~WorkerFunction
+     * @function
+     * @returns {boolean|string|SDP} there are several things that this
+     * function can return:
+     * <tt>true</tt> if the renegotiation should follow the function execution
+     * <tt>false</tt> if there should be no renegotiation
+     * <tt>SDP</tt> that will be set as the remote description in the
+     * renegotiation process
+     * <tt>string</tt> which stands for the error description, no renegotiation
+     * will be done in that case it's returned
      */
-    removeStreamFromPeerConnection(stream) {
-        const actualStream
-            = stream && stream.getOriginalStream
-                ? stream.getOriginalStream() : stream;
+    /**
+     * Does the logic of doing the session renegotiation by updating local and
+     * remote session descriptions. Will compare the local description, before
+     * and after the renegotiation to update local streams description (sends
+     * "source-add"/"source-remove" notifications).
+     * @param {string} actionName the name of the action which will appear in
+     * the events logged to the logger.
+     * @param {WorkerFunction} workFunction a function that will be executed on
+     * the queue. See type description for moe info.
+     * @private
+     */
+    _doRenegotiate(actionName, workFunction) {
```

Comment:
this seems to have gotten a lot more complicated (with multiple layers of work functions and wrappers) and it's not clear to my why it has to be that way.  maybe we can talk about that.